sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala

```diff
@@ -2336,46 +2336,40 @@ class DataFrameSuite extends QueryTest with SharedSQLContext {
 
     val sourceDF = spark.createDataFrame(rows, schema)
 
-    val structWhenDF = sourceDF
+    def structWhenDF: DataFrame = sourceDF
       .select(when('cond, struct(lit("a").as("val1"), lit(10).as("val2"))).otherwise('s) as "res")
       .select('res.getField("val1"))
-    val arrayWhenDF = sourceDF
+    def arrayWhenDF: DataFrame = sourceDF
       .select(when('cond, array(lit("a"), lit("b"))).otherwise('a) as "res")
       .select('res.getItem(0))
-    val mapWhenDF = sourceDF
+    def mapWhenDF: DataFrame = sourceDF
       .select(when('cond, map(lit(0), lit("a"))).otherwise('m) as "res")
       .select('res.getItem(0))
 
-    val structIfDF = sourceDF
+    def structIfDF: DataFrame = sourceDF
       .select(expr("if(cond, struct('a' as val1, 10 as val2), s)") as "res")
       .select('res.getField("val1"))
-    val arrayIfDF = sourceDF
+    def arrayIfDF: DataFrame = sourceDF
       .select(expr("if(cond, array('a', 'b'), a)") as "res")
       .select('res.getItem(0))
-    val mapIfDF = sourceDF
+    def mapIfDF: DataFrame = sourceDF
       .select(expr("if(cond, map(0, 'a'), m)") as "res")
       .select('res.getItem(0))
 
-    def checkResult(df: DataFrame, codegenExpected: Boolean): Unit = {
-      assert(df.queryExecution.executedPlan.isInstanceOf[WholeStageCodegenExec] == codegenExpected)
-      checkAnswer(df, Seq(Row("a"), Row(null)))
+    def checkResult(): Unit = {
+      checkAnswer(structWhenDF, Seq(Row("a"), Row(null)))
+      checkAnswer(arrayWhenDF, Seq(Row("a"), Row(null)))
+      checkAnswer(mapWhenDF, Seq(Row("a"), Row(null)))
+      checkAnswer(structIfDF, Seq(Row("a"), Row(null)))
+      checkAnswer(arrayIfDF, Seq(Row("a"), Row(null)))
+      checkAnswer(mapIfDF, Seq(Row("a"), Row(null)))
     }
 
-    // without codegen
-    checkResult(structWhenDF, false)
-    checkResult(arrayWhenDF, false)
-    checkResult(mapWhenDF, false)
-    checkResult(structIfDF, false)
-    checkResult(arrayIfDF, false)
-    checkResult(mapIfDF, false)
-
-    // with codegen
-    checkResult(structWhenDF.filter('cond.isNotNull), true)
```

Comment:
no matter we add filter or not, the `Project` will always be evaluated without codegen, because it's above local relation  and the optimizer will evaluate it eagerly.

Reply:
If it's the case why the [assert](https://github.com/apache/spark/pull/21795/files/d1bc6124628808444a4ab5bed6bfc6f897d76e03#diff-5d2ebf4e9ca5a990136b276859769289L2360) didn't fail?

Reply:
hmm, this is the execution plan I see for `structWhenDF.filter('cond.isNotNull)`:

```scala
*(1) Project [CASE WHEN cond#77042 THEN [a,10] ELSE s#77043 END.val1 AS res.val1#77054]
+- *(1) Filter isnotnull(cond#77042)
   +- LocalTableScan [cond#77042, s#77043, a#77044, m#77045]
```

Reply:
This is the execution plan for `structWhenDF`:
```scala
LocalTableScan [res.val1#77079]      
```

Reply:
ah that's tricky. Because filter pushdown runs first, the local relation optimization can't be applied.

To prevent confusions like this, how about we use local/cached relation to test it?

Reply:
I saw some tests using similar dummy filters in `DataFrameFunctionsSuite`. Should we fix them as well?

Reply:
@cloud-fan Thanks for the clarification and this PR!

Btw, there are many tests in ```DataFrameFunctionsSuite``` that test only the scenarios without codgen. WDYT about adding a generic ```checkAnswer``` method to ```QueryTest``` that would evaluate a dataframe for both cases similarly like ```ExressionEvalHelper.checkEvaluation``` does for expressions? If it's possible, of course.

Reply:
it will be very hard to write a general `checkAnswer`, because the local relation optimization can only handle `Project`. I'd like to wait for the general codegen config.